### PR TITLE
Files: Extend search to also cover tags

### LINKED ISSUE #326: Use the search bar to search for files by tag
Currently to find files/folders with a particular tag within NextCloud you need to go to the "Tags" menu option then type in or select the tag(s) that you want to filter with.

Using the search function, it is not possible to search for files/folders with a specified tag as it does a full text search of the files, but not the associated tag.

Can the search function be enhanced to search for not only the full text/file names but also for the tag(s) associated with a file?  Another possible enhancement of the search can be to search for keywords entered in the search function within the comments associated with a file/folder

### PR DESCRIPTION
**Pull request by marcelklehr**
fixes #326

This is my first real server PR so please bear with me :)

### CODE REVIEW

**Comment by Spartachetto** [comment]
@marcelklehr thank you!!!!!!!!!!
That is a feature request from **2016**!

Beyond that, I'd like to remember two other issues:

- #4851 (Possibility to search system tags with a better query system)
- #2143 (Visualize useful synthetic information when user clicks 'Tags' )

This just because maybe it could be useful to review this code keeping in mind the requirements of those FR. 
Thanks again.

File: apps/files/lib/Search/FilesSearchProvider.php [hunk 1]
```diff
@@ -103,7 +104,11 @@ public function getOrder(string $route, array $routeParameters): int {
 	public function search(IUser $user, ISearchQuery $query): SearchResult {
 		$userFolder = $this->rootFolder->getUserFolder($user->getUID());
 		$fileQuery = new SearchQuery(
-			new SearchComparison(ISearchComparison::COMPARE_LIKE, 'name', '%' . $query->getTerm() . '%'),
```
Comment (marcelklehr) [inline apps/files/lib/Search/FilesSearchProvider.php:106]: by the way, was it intentional to not escape % and _ in the search term here? It's a nice feature to be able to search for files with %, but I'm not sure that was intended

**Comment by marcelklehr** [comment]
```
Failed scenarios:
 /drone/src/tests/acceptance/features/app-files-tags.feature:22
```

But I can't replicate that manually.

**Comment by icewind1991** [comment]
I'm a bit concerned about the performance impact of joining the tag tables for every query

**Comment by marcelklehr** [comment]
> I'm a bit concerned about the performance impact of joining the tag tables for every query

The systemtag and vcategory tables are only joined when the query involves the fields `tagname` or `systemtag`. Which for the FilesSearchProvider is indeed every time.

Can we extract something actionable from this concern? Should I run a benchmark? Should we cc someone else to weigh in? Should we close the feature request as wontfix due to performance reasons?

**Comment by bonswouar** [comment]
> Can we extract something actionable from this concern? Should I run a benchmark? Should we cc someone else to weigh in? Should we close the feature request as wontfix due to performance reasons?

Any news about this? @marcelklehr How's the performance impact according to your tests?

If I may suggest,
- Maybe we could make tag search optional in the settings at first, informing this is an experimental feature that could slow down search. I'm not sure this is a very nextcloud-way to do though.
- OR why not add a prefix to be able to search tags, like the usual `#` (or the Google way using a full keyword prefix, like `tag:`)? Thus we could join those columns only when needed.

Anyway fixing this (quite old) issue would allow to continue on this subject, have a better look at other related issues like #4851 and look for some way of optimizing search afterwards!

**Comment by marcelklehr** [comment]
> How's the performance impact according to your tests?

I haven't tested performance so far.

Making it optional doesn't sound like a good solution and adding \#-based tags search would only be an extension to this PR, IMHO.

**Comment by icewind1991** [comment]
Also, I think it would be better to separate out the results for search by tag from the results by filename, adding it as a separate search provider.

This both makes it more clear why a search results shows up and prevents one result type to drown out the other type

**Comment by jancborchardt** [comment]
> > This both makes it more clear why a search results shows up and prevents one result type to drown out the other type
> 
> I agree that it should be clear why a search result is included, but I'm not sure adding a separate search provider is the way to go.
> 
> cc @nextcloud/designers ?

Sorry @marcelklehr! I thought I replied to this here or via chat already. UI-wise I would definitely keep it in the same section as "Files", since tags are just as related to the file like filename, file contents, etc. – shouldn’t have a separate section, cause you are looking for Files with that tag, not for the Tags. :)

**Comment by marcelklehr** [comment]
So, clearly we are at an impasse here. UX-wise it would be nice to be able to search for files based on tags as part of the normal file search, while performance-wise it's likely more optimal to not include tags in the file search. I would love to see this PR get merged, but to me it's not clear what I need to do for that.

File: lib/private/Files/Cache/SearchBuilder.php [hunk 2]
```diff
@@ -162,8 +162,12 @@ private function getOperatorFieldAndValue(ISearchComparison $operator) {
 		} elseif ($field === 'favorite') {
 			$field = 'tag.category';
 			$value = self::TAG_FAVORITE;
+		} elseif ($field === 'name') {
+			$field = 'file.name';
 		} elseif ($field === 'tagname') {
 			$field = 'tag.category';
+		} elseif ($field === 'systemntag') {
```
Comment (szaimen) [inline lib/private/Files/Cache/SearchBuilder.php:169]: typo?

**Review by szaimen — COMMENTED** [review]
@marcelklehr question: what happens if the systemtags app is disabled? will it bug out?

**Comment by marcelklehr** [comment]
Testing with my photo collection, a query with tags took 400ms while a search without took ~100ms :/

**Comment by marcelklehr** [comment]
A different imeplementation approach would be to avoid the database joins and extend the file cache table with a direct tags field, so search would be faster. What do you think @icewind1991 ?

**Comment by marcelklehr** [comment]
So, I have multiple proposals on how to proceed:

1. avoid the database joins and extend the extended file cache table with a direct tags field, so search is faster
2. Only enable tag search when systemtags are enabled

cc @jancborchardt @icewind1991 @szaimen

**Comment by szaimen** [comment]
> So, I have multiple proposals on how to proceed:
> 
> 1. avoid the database joins and extend the extended file cache table with a direct tags field, so search is faster
> 2. Only enable tag search when systemtags are enabled
> 
> cc @jancborchardt @icewind1991 @szaimen

I'd go with a config option. Would that be fine for @jancborchardt and @icewind1991 and @skjnldsv ?

**Comment by nimishavijay** [comment]
I think tag search would be super useful (considering how many issues and PRs are centered around it) so it should be enabled by default :)

> 1. avoid the database joins and extend the extended file cache table with a direct tags field, so search is faster
> 2. Only enable tag search when systemtags are enabled

Not sure about how difficult this is from a technical POV, but ideally we should make it work faster :) I am assuming that any sort of tags would not work on your NC instance if systemtags is not enabled (is that correct?) so then tag search would not be useful anyway, right?

> I'd go with a config option. 

Config option to turn disable this feature as a last resort in case it is not fast enough sounds like a good idea to me :)

**Comment by icewind1991** [comment]
> avoid the database joins and extend the extended file cache table with a direct tags field, so search is faster

unsure if that would really help but might be worth a try.

Another approach might be to first check if there are any tags that match the search term (should be fairly cheap since it's unlikely that there are that many tags?) and then get the files by tagid only if some tags are found.

--

In general though I'm not entirely sure if searching files by tag name is actually useful when done this way.
Once a tag contains more than a dozen files, searching for the tag will just always return the results with no real way of filtering down the results any further.

Instead it seems more useful to have the search return the actual tags instead of just the tagged files, maybe including a small sample of tagged files and a link to the relevant tag in the files app.

Quick mockup:

mockup of tag search

This solves the problem of a single large tag completely dominating the search results, guides the user to a place that is better suited to finding files in a large tag and leaves the search-by-name useful as is.

**Comment by nimishavijay** [comment]
Looks good! 2 small suggestions:
- reduce the size of the icon a bit, and its color should be `var(--color-main-text)`
- show the tag that you have searched for first in the list of tags below the file
so it would be "nature, landscape" and "nature, alpine, landscape, mountain"

Great work! :)

**Comment by marcelklehr** [comment]
Thanks for the review!

> reduce the size of the icon a bit, and its color should be var(--color-main-text)

Not sure if that's possible as I didn't write the code for that popover, I'm just passing the icon name to the Search API.

> show the tag that you have searched for first in the list of tags below the file

Ah, nice, will do :)

**Comment by marcelklehr** [comment]
Peek 2022-05-10 11-42

**Comment by marcelklehr** [comment]
@szaimen How do I make node CI pass?

**Review by icewind1991 — APPROVED** [review]
Room for improvement with some frontend changes but works as is

**Comment by jancborchardt** [comment]
Nice @marcelklehr! Additional feedback:
- How about first showing the files and the tag entry below. Wording for that could be "All tagged nature …"
- The subline wording, instead of "with landscape, nature", more readable and accessible could be "tagged landscape, nature"?

**Comment by marcelklehr** [comment]
@jancborchardt Thanks for the feedback! :)

Peek 2022-05-16 12-34

(Don't be distracted by the cute owl...)

**Review by jancborchardt — APPROVED** [review]
Looks great! And very cute owl indeed :D

**Comment by marcelklehr** [comment]
Not sure why S3 is failing, seems unrelated.

**Comment by szaimen** [comment]
@marcelklehr can you please resolve DCO? Afterwards I think we can merge this :)

**Comment by marcelklehr** [comment]
Done :)

**Comment by szaimen** [comment]
> Done :)

Thanks! I guess squashing would also be good, then I merge :)

**Comment by marcelklehr** [comment]
Done :)

**Comment by bonswouar** [comment]
Thank you so much @marcelklehr for this, and Nextcloud's maintainers for your feedbacks and accepting to merge it! This feature will make a big difference, love it!
_(I know this isn't a very useful comment but I had to communicate my appreciation ;))_

**Comment by szaimen** [comment]
Thanks! CI failure unrelated

**Comment by BloodyIron** [comment]
yay! \o/

**Comment by darguez** [comment]
Is it posible to select the searched files (by tag) to perform any action with them? (f.i. copy them to a folder to share them). Why the search view does not show the checkboxes? Thanks and great work!

**Comment by nkeilar** [comment]
Looks like a great improvement - good work. Just one small issue - it doesn't seem working for me.  

We are running 24.0.5 but when I search for folder with tags, no results are showing. Does anything need to be done to enable this functionality? 

Additionally, we actually really want to be able to use an API to query NC for a list of files/folders with a given tag. This is for integration with a third parry app we are building. Is their a search API that a client could use to search NC by tags? 

Thanks in advance for any assistance - we are a bit stuck!

**Comment by tcitworld** [comment]
> We are running 24.0.5 but when I search for folder with tags, no results are showing. Does anything need to be done to enable this functionality?

As per the milestone set on the right sidebar of this page, this is a Nextcloud 25 feature.

**Comment by NablaMetal** [comment]
Sorry to ask that, but I can't get the search bar to search for tags (tags filtering in left pane of file app works).
Is there any particular thing to configure (like in config.php, or admin panel ?)
NC 27.1.1

**Comment by ChristophWurst** [comment]
@NablaMetal please use https://help.nextcloud.com/ for support :pray:

**Comment by BloodyIron** [comment]
Not seeing this mentioned in the NC 25.0.0 release announcement.. What version did this actually make it into? : https://nextcloud.com/blog/announcing-nextcloud-hub-3-brand-new-design-and-photos-2-0-with-editor-and-ai/

**Comment by marcelklehr** [comment]
> This pull request first appeared in [v25.0.0beta1](https://github.com/nextcloud/server/releases/tag/v25.0.0beta1)

@BloodyIron v25

**Comment by BloodyIron** [comment]
> > This pull request first appeared in [v25.0.0beta1](https://github.com/nextcloud/server/releases/tag/v25.0.0beta1)
> 
> @BloodyIron v25

I don't see release notes referencing this issue# or commit, and checking the changes between v25.0.0beta1 and v24.0.0 is over 1200 commits. So I don't see a way to explicitly confirm it. And while I suspect you're right, I'm a fan of a bird in the hand. Unless there's a method I'm not seeing? 🤔

**Comment by marcelklehr** [comment]
@BloodyIron You can view the commit in this PR and github will tell you which tags the commit appears in. The first tag is something with v25

**Pull request merged**